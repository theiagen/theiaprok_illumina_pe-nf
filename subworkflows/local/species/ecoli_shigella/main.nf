include { STXTYPER        } from '../../../../modules/local/species_typing/escherichia_shigella/stxtyper/main'
include { ECTYPER         } from '../../../../modules/local/species_typing/escherichia_shigella/ectyper/main'
include { SHIGATYPER      } from '../../../../modules/local/species_typing/escherichia_shigella/shigatyper/main'
include { SHIGEIFINDER    } from '../../../../modules/local/species_typing/escherichia_shigella/shigeifinder/main'
include { VIRULENCEFINDER } from '../../../../modules/local/species_typing/escherichia_shigella/virulencefinder/main'
include { SONNEITYPER     } from '../../../../modules/local/species_typing/escherichia_shigella/sonneityping/main'
include { SEROTYPEFINDER  } from '../../../../modules/local/species_typing/escherichia_shigella/serotypefinder/main'

workflow ESCHERICHIA_SHIGELLA_TYPING {

    take:
    ch_ecoli_shigella // Channel of tuples [meta, assembly, reads, species]


    main:

    ch_by_species = ch_ecoli_shigella.branch {
        ecoli_shigella: it[3] == "Escherichia" || it[3] == "Shigella sonnei" || it[3] == "Escherichia coli"
        sonnei_only: it[3] == "Shigella sonnei" && !params.assembly_only
        shiga_typer: it[3] == "Shigella sonnei" && !params.assembly_only && params.paired_end && !params.ont_data
    }

    ch_assembly = ch_by_species.ecoli_shigella.map { meta, assembly, reads, species ->
        [meta, assembly]
    }
    ch_sonneityper = ch_by_species.sonnei_only.map { meta, assembly, reads, species ->
        [meta, reads]
    }
    ch_shigatyper = ch_by_species.shiga_typer.map { meta, assembly, reads, species ->
        [meta, reads]
    }

    ch_versions = Channel.empty()
    ch_stxtyper_results = Channel.empty()
    ch_serotypefinder_results = Channel.empty()
    ch_ectyper_results = Channel.empty()
    ch_shigeifinder_results = Channel.empty()
    ch_virulencefinder_results = Channel.empty()
    ch_shigatyper_results = Channel.empty()
    ch_sonneityper_results = Channel.empty()
    ch_value_results = Channel.empty()

    // STX typer - special case: auto-run on Escherichia/Shigella, optional on others
    if (params.call_stxtyper) {
        STXTYPER (ch_assembly)
        ch_value_results = ch_value_results.mix(STXTYPER.out.stxtyper_value_results)
        ch_stxtyper_results = STXTYPER.out.stxtyper_report
        ch_versions = ch_versions.mix(STXTYPER.out.stxtyper_version)
    }

    SEROTYPEFINDER (
        ch_assembly
    )
    ch_value_results = ch_value_results.mix(SEROTYPEFINDER.out.serotypefinder_serotype)
    ch_serotypefinder_results = SEROTYPEFINDER.out.serotypefinder_report
    ch_versions = ch_versions.mix(SEROTYPEFINDER.out.versions)
        
    ECTYPER (
        ch_assembly,
        params.ectyper_o_min_percent_identity ?: 90,
        params.ectyper_h_min_percent_identity ?: 95,
        params.ectyper_o_min_percent_coverage ?: 90,
        params.ectyper_h_min_percent_coverage ?: 50,
        params.ectyper_verify ?: false,
        params.ectyper_print_alleles ?: false
    )
    ch_value_results = ch_value_results.mix(ECTYPER.out.ectyper_predicted_serotype_file)
    ch_ectyper_results = ECTYPER.out.ectyper_results
    ch_versions = ch_versions.mix(ECTYPER.out.versions)

    SHIGEIFINDER (
        ch_assembly
    )
    ch_value_results = ch_value_results.mix(SHIGEIFINDER.out.shigeifinder_value_results)
    ch_shigeifinder_results = SHIGEIFINDER.out.shigeifinder_report
    ch_versions = ch_versions.mix(SHIGEIFINDER.out.versions)

    VIRULENCEFINDER (
        ch_assembly,
        params.virulencefinder_database ?: "virulence_ecoli",
        params.virulencefinder_min_percent_coverage ?: 0.60,
        params.virulencefinder_min_percent_identity ?: 0.80
    )
    ch_value_results = ch_value_results.mix(VIRULENCEFINDER.out.virulence_factors)
    ch_virulencefinder_results = VIRULENCEFINDER.out.virulence_report
    ch_versions = ch_versions.mix(VIRULENCEFINDER.out.versions)

    SHIGATYPER (ch_shigatyper,params.ont_data ?: false)
    ch_value_results = ch_value_results.mix(SHIGATYPER.out.shigatyper_value_results)
    ch_shigatyper_results = SHIGATYPER.out.shigatyper_summary
    ch_versions = ch_versions.mix(SHIGATYPER.out.versions)

    SONNEITYPER (
        ch_sonneityper,
        params.ont_data ?: false
    )
    ch_value_results = ch_value_results.mix(SONNEITYPER.out.sonneityping_value_results)
    ch_sonneityper_results = SONNEITYPER.out.sonneityping_final_report
    ch_versions = ch_versions.mix(SONNEITYPER.out.versions)

    emit:
    stxtyper_results = ch_stxtyper_results
    serotypefinder_results = ch_serotypefinder_results
    ectyper_results = ch_ectyper_results
    shigeifinder_results = ch_shigeifinder_results
    virulencefinder_results = ch_virulencefinder_results
    shigatyper_results = ch_shigatyper_results
    sonneityper_results = ch_sonneityper_results
    value_results = ch_value_results
    versions = ch_versions
}
