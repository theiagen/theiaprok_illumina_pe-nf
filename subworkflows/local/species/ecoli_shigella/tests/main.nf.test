nextflow_workflow {

    name "Test Subworkflow ESCHERICHIA_SHIGELLA_TYPING"
    script "../main.nf"
    workflow "ESCHERICHIA_SHIGELLA_TYPING"

    // Currently utilizes local test data, needs to be replaced with remote data
    test("escherichia_shigella - characterization") {
        when {
            params {
                // Use minimal parameters for testing
                call_stxtyper = true
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_ecoli-shigella', single_end:false ],
                    file("gs://fc-c37b6bc6-b033-4ef1-b8aa-b8c1d25524d4/submissions/9b83240f-3549-42f5-8960-d6d04ed6555b/theiaprok_illumina_pe/88ab6224-07c8-451c-9d75-c440f4b89b6e/call-digger_denovo/digger_denovo/424e44ca-ecad-4652-aefd-f2405b33d086/call-filter_contigs/cacheCopy/19050801924_filtered_contigs.fasta"),
                    [
                        file("gs://theiagen-large-public-files-rp/terra/theiaprok-files/theiaprok_validation/19050801924-OR-M05181-200429_S10_L001_R1_001.fastq.gz"),
                        file("gs://theiagen-large-public-files-rp/terra/theiaprok-files/theiaprok_validation/19050801924-OR-M05181-200429_S10_L001_R2_001.fastq.gz")
                    ],
                    "Escherichia coli"
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.stxtyper_results,
                    workflow.out.serotypefinder_results,
                    workflow.out.ectyper_results,
                    workflow.out.shigeifinder_results,
                    workflow.out.virulencefinder_results,
                    workflow.out.shigatyper_results,
                    workflow.out.sonneityper_results,
                    workflow.out.versions).match() }
            )
        }
    }
        test("escherichia_shigella - characterization - no stxtyper") {
        when {
            params {
                // Use minimal parameters for testing
                call_stxtyper = false
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_ecoli-shigella', single_end:false ],
                    file("gs://fc-c37b6bc6-b033-4ef1-b8aa-b8c1d25524d4/submissions/9b83240f-3549-42f5-8960-d6d04ed6555b/theiaprok_illumina_pe/88ab6224-07c8-451c-9d75-c440f4b89b6e/call-digger_denovo/digger_denovo/424e44ca-ecad-4652-aefd-f2405b33d086/call-filter_contigs/cacheCopy/19050801924_filtered_contigs.fasta"),
                    [
                        file("gs://theiagen-large-public-files-rp/terra/theiaprok-files/theiaprok_validation/19050801924-OR-M05181-200429_S10_L001_R1_001.fastq.gz"),
                        file("gs://theiagen-large-public-files-rp/terra/theiaprok-files/theiaprok_validation/19050801924-OR-M05181-200429_S10_L001_R2_001.fastq.gz")
                    ],
                    "Escherichia coli"
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.serotypefinder_results,
                    workflow.out.ectyper_results,
                    workflow.out.shigeifinder_results,
                    workflow.out.virulencefinder_results,
                    workflow.out.shigatyper_results,
                    workflow.out.sonneityper_results,
                    workflow.out.versions).match() }
            )
        }
    }
}