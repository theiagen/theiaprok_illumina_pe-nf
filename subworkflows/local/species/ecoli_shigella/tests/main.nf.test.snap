{
    "escherichia_shigella - characterization - no stxtyper": {
        "content": [
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella_results_tab.tsv:md5,1302b591d230541196a3f04b499ac647"
                ]
            ],
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella.tsv:md5,cac1ccf09d4bdbffa00c66625f4c966a"
                ]
            ],
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella_shigeifinder.tsv:md5,f3216127bb1bba962d84f98fe0042390"
                ]
            ],
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella_results_tab.tsv:md5,676b7d5edfe724d7df0b48001a9d11a1"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,200f918ddf823e91273ac112dc806d70",
                "versions.yml:md5,c4d679f7b5691fc311aaf2f7965475be",
                "versions.yml:md5,c5e01602e6325888201c1d8f5e616480",
                "versions.yml:md5,d95a9584be8bcc5757f3f6b3342229cb"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-07-02T20:07:03.681751175"
    },
    "escherichia_shigella - characterization": {
        "content": [
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella_stxtyper.tsv:md5,61f9260cfa9779f36d28ee83e80f4a32"
                ]
            ],
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella_results_tab.tsv:md5,1302b591d230541196a3f04b499ac647"
                ]
            ],
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella.tsv:md5,cac1ccf09d4bdbffa00c66625f4c966a"
                ]
            ],
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella_shigeifinder.tsv:md5,f3216127bb1bba962d84f98fe0042390"
                ]
            ],
            [
                [
                    {
                        "id": "test_ecoli-shigella",
                        "single_end": false
                    },
                    "test_ecoli-shigella_results_tab.tsv:md5,676b7d5edfe724d7df0b48001a9d11a1"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,200f918ddf823e91273ac112dc806d70",
                "versions.yml:md5,419f94e9e74f7e054849fbddb0bb4530",
                "versions.yml:md5,c4d679f7b5691fc311aaf2f7965475be",
                "versions.yml:md5,c5e01602e6325888201c1d8f5e616480",
                "versions.yml:md5,d95a9584be8bcc5757f3f6b3342229cb"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-07-02T20:06:52.532568785"
    }
}